git-blame: edit page (#20768)
tldr-pages/tldr: pages/common/git-blame.md

# git blame

> Show what commit and author last modified each line of a file.
> More information: <https://git-scm.com/docs/git-blame>.

- Print a file with authorship info (author name and commit hash):

`git blame {{path/to/file}}`

- Print a file with authorship info (author email and commit hash):

`git blame {{[-e|--show-email]}} {{path/to/file}}`

- Print a file with authorship info as of a specific commit:

`git blame {{commit}} {{path/to/file}}`

- Print a file with authorship info before a specific commit:

`git blame {{commit}}~ {{path/to/file}}`

- Print a file with authorship info starting at a given line:

> `git blame -L {{123}}`

Author: Oh wow, great catch! :facepalm:

- Annotate a specific line range of a file:

`git blame -L {{start_line}},{{end_line}} {{path/to/file}}`

- Annotate 10 lines of a file starting at the first line matching a given text:

> `git blame -L '/{{text}}/',+10 {{path/to/file}}`

Reviewer: Why can't we be both didactic and provide a useful complete command? The previous examples already provide pretty good basis for the how the command works in simple terms. Since people are used to the way you can chase git blame in github, to me it feels correct to provide the same functionality in the command line at the end of the page.

> `git blame -L '/{{text}}/',+10 {{path/to/file}}`

Author: But isn't GitHub a behavior already supported by the previous example? 🤔

Why overload this example about jumping to a line with a text pattern + a line offset (already a relatively complex combination which is only ok because we already introduce the range syntax in a previous example) with yet an extra optional argument?

Doing so would not just make this more complex than necessary, but would also make the description either too verbose or incomplete.

> `git blame -L '/{{text}}/',+10 {{path/to/file}}`

Author: > Sort of yes, but when I've tried them in practice, they fail because the lines that specific code exists in moves as code gets added or removed above it.

Exactly — the GitHub experience is based on line numbers, not text matches. And besides, the text you're tracing is just as likely to change as the line numbers where the text appears are to jump around (considering that, if you're interested in a specific section of the code, you'll likely be looking at the commits that do modify that section).

> This is fine in my opinion since the simple versions of the commands are already covered by the previous examples.

I don't see tldr-pages entries as a set of simple examples followed by a set of complex ones, but as a cohesive unit taking the reader through a journey of learning how the command works. Having simpler examples earlier in the page does not IMO justify dropping the goal of clarity and progressive disclosure (adding one new thing at a time) in the later ones.

> I don't think - Jump to the parent of a specific commit and track a specific text and 10 of its following lines: is too complex

It is not too complex, sure, but it is _more_ complex, both linguistically and conceptually (since the example will be mixing up more concepts simultaneously). I feel we can avoid that while still providing the reader with the knowledge needed to build up this example themselves — we don't have to serve the example ready-made.

> Oh on a side note, what's the difference between `commit^` and `commit~`?

Yeah, that's a nice detail of git that often gets overlooked. The former is to select (parallel) parents of a merge commit, which can have two or more parents (`^` = `^1` = the first parent; `^2` = the second parent; `^3` = the third parent, etc.). The latter is for selecting (sequential) ancestors of a commit, i.e. `~` = `~1` = the parent; `~2` = the grandparent; `~3` = the great-grand-parent, and so on.

# git blame

> Print a file annotating each line with authorship data

> > (e.g. hash and author) from the last commit that modified it.

Author: Git's own docs have a shorter and equally clear description for the command:

> Show what revision and author last modified each line of a file

WDYT about reusing it? (Btw, if we do, I'd argue for using "commit" instead of "revision" — the latter is technically correct, but someone who already knows git will know that "commit" can be replaced with "revision", but a novice who has only heard of commits won't know the opposite. So my proposal would be:

```suggestion
> Show what commit and author last modified each line of a file.
```
